File: Year-2023/LeetCode-August-Challenge/NumberofMusicPlaylists.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
private:
    long long MOD = 1e9 + 7;
    long long helper(int n, int goal, int k, vector<vector<long long>> &dp){
        if(n == 0 && goal == 0) return 1;
        if(n == 0 || goal == 0) return 0;
        if(dp[n][goal] != -1) return dp[n][goal];

        long long playOld = helper(n, goal - 1, k, dp) * (max(0, n - k));
        long long playNew = helper(n - 1, goal - 1, k, dp) * n;

        return dp[n][goal] = (playOld + playNew) % MOD;
    }
public:
    int numMusicPlaylists(int n, int goal, int k) {
        vector<vector<long long>> dp(n + 1, vector<long long>(goal + 1, -1));
        return helper(n, goal, k, dp);
    }
};

int main(){
    int n = 3, goal = 3, k = 1;
    Solution *s = new Solution();
    cout<<s->numMusicPlaylists(n, goal, k)<<endl;
    return 0;
}

File: Year-2023/LeetCode-August-Challenge/NumberofMusicPlaylists.java
class Solution {
    long mod = (long) 1e9 + 7;

    public int numMusicPlaylists(int N, int L, int K) {
        long[][] dp = new long[N + 1][L + 1];
        for (int i = K + 1; i <= N; ++i)
            for (int j = i; j <= L; ++j)
                if ((i == j) || (i == K + 1))
                    dp[i][j] = factorial(i);
                else
                    dp[i][j] = (dp[i - 1][j - 1] * i + dp[i][j - 1] * (i - K)) % mod;
        return (int) dp[N][L];
    }

    long factorial(int n) {
        return n > 0 ? factorial(n - 1) * n % mod : 1;
    }
}

public class NumberofMusicPlaylists {
    public static void main(String[] args) {
        int n = 3, goal = 3, k = 1;
        System.out.println(new Solution().numMusicPlaylists(n, goal, k));
    }
}